ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PrintPublisher.java

```diff
@@ -0,0 +1,136 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.rest.server.resources.streaming;
+
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.BytesDeserializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.connect.data.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.ksql.rest.server.resources.streaming.PollingSubscription.Pollable;
+import io.confluent.ksql.rest.server.resources.streaming.TopicStream.RecordFormatter;
+
+public class PrintPublisher implements Flow.Publisher<Collection<String>> {
+
+  private static final Logger log = LoggerFactory.getLogger(PrintPublisher.class);
+
+  private final ListeningScheduledExecutorService exec;
+  private final SchemaRegistryClient schemaRegistryClient;
+  private final String topicName;
+  private final boolean fromBeginning;
+  private final Map<String, Object> consumerProperties;
+
+  private boolean closed = false;
+  private KafkaConsumer<String, Bytes> topicConsumer;
+
+  public PrintPublisher(
+      ListeningScheduledExecutorService exec,
+      SchemaRegistryClient schemaRegistryClient,
+      Map<String, Object> consumerProperties,
+      String topicName,
+      boolean fromBeginning
+  ) {
+    this.exec = exec;
+    this.schemaRegistryClient = schemaRegistryClient;
+    this.consumerProperties = consumerProperties;
+    this.topicName = topicName;
+    this.fromBeginning = fromBeginning;
+  }
+
+  @Override
+  public void subscribe(Flow.Subscriber<Collection<String>> subscriber) {
+    this.topicConsumer = new KafkaConsumer<>(
+        consumerProperties,
+        new StringDeserializer(),
+        new BytesDeserializer()
+    );
+
+    log.info("Running consumer for topic {}", topicName);
+    List<TopicPartition> topicPartitions = topicConsumer.partitionsFor(topicName)
+        .stream()
+        .map(partitionInfo -> new TopicPartition(partitionInfo.topic(), partitionInfo.partition()))
+        .collect(Collectors.toList());
+    topicConsumer.assign(topicPartitions);
+
+    if (fromBeginning) {
+      topicConsumer.seekToBeginning(topicPartitions);
+    }
+    subscriber.onSubscribe(new PollingSubscription<>(
+        exec,
+        subscriber,
+        new Pollable<Collection<String>>() {
```

Comment:
how would you suggest to do that?

Reply:
If you make `Pollable#poll(..)` return something like a `PollableResult<T>` where `PollableResult` it might look like:
```
public static class PollableResult<T> {
    private final Optional<T> result;
    private final Optional<Throwable> exception;

    public PollableResult() {
      this((T) null);

    }

    public PollableResult(final T result) {
      this.result = Optional.ofNullable(result);
      this.exception = Optional.empty();
    }

    public PollableResult(final Throwable e) {
      this.result = Optional.empty();
      this.exception = Optional.of(e);
    }

    public boolean isError() {
      return exception.isPresent();
    }

    public Throwable error() {
      return exception.orElse(null);
    }

  }
```

Add helper methods to create empty results etc.

Then in `PollingSubscription` you can get rid of the `draining` flag and check the status of the result instead, so it might end up being something like:

```
@Override
  public void request(long n) {
    // check draining status since request can be reentrant
    if (!result.isError()) {
      future = exec.submit(() -> {
        result = pollable.poll();
        if (!result.result.isPresent()) {
          future = exec.schedule(() -> request(1), BACKOFF_DELAY_MS, TimeUnit.MILLISECONDS);
        } else {
          subscriber.onNext(result.result.get());
        }
        if (result.isError()) {
          pollable.close();
          subscriber.onError(result.error());
        }
      });
    }
  }
```

Then I believe the `Pollable` interface can be reduced to:
```
public interface Pollable<T> {

    PollableResult<T> poll();

    void close();
  }
```

Reply:
@xvrl `PollableResult` is written once and not for every subscription, so there is some value in that. Granted it may be too early to tell as we only have two implementations.
As for the draining bit, i don't really follow what you are saying. As it is now it seems that once the `draining` flag has been set then that block of code will no longer be executed. What am i missing? This is w.here tests would be useful...

Reply:
@dugy added tests for all the critical sections, and rewrote to remove `Pollable` altogether